components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/AbstractUserInfoResponseBuilder.java

```diff
@@ -0,0 +1,203 @@
+
+        // Handle subject claim.
+        String subjectClaim = getSubjectClaim(userClaims, spTenantDomain, tokenResponse);
+        filteredUserClaims.put(OAuth2Util.SUB, subjectClaim);
+
+        // Handle essential claims
+        Map<String, Object> essentialClaims = getEssentialClaims(tokenResponse, userClaims);
+        filteredUserClaims.putAll(essentialClaims);
+
+        return buildResponse(tokenResponse, spTenantDomain, filteredUserClaims);
+    }
+
+
+    protected String getSubjectClaim(Map<String, Object> userClaims,
+                                     String spTenantDomain,
+                                     OAuth2TokenValidationResponseDTO tokenResponse) throws UserInfoEndpointException {
+
+        String subjectClaim = (String) userClaims.get(OAuth2Util.SUB);
+        if (StringUtils.isBlank(subjectClaim)) {
+            // We need to send the tenant aware and user store domain removed name.
+            String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(tokenResponse.getAuthorizedUser());
+            subjectClaim = UserCoreUtil.removeDomainFromName(tenantAwareUsername);
+        }
+        return returnSubjectClaim(subjectClaim, spTenantDomain, tokenResponse);
+    }
+
+    private String returnSubjectClaim(String sub,
+                                      String tenantDomain,
+                                      OAuth2TokenValidationResponseDTO tokenResponse) throws UserInfoEndpointException {
+
+        String clientId = getClientId(tokenResponse);
+        ServiceProvider serviceProvider = getServiceProvider(tenantDomain, clientId);
+
+        String userTenantDomain = MultitenantUtils.getTenantDomain(tokenResponse.getAuthorizedUser());
+        String userStoreDomain = UserCoreUtil.extractDomainFromName(tokenResponse.getAuthorizedUser());
+
+        if (serviceProvider != null) {
+            boolean isUseTenantDomainInLocalSubject = serviceProvider.getLocalAndOutBoundAuthenticationConfig()
+                    .isUseTenantDomainInLocalSubjectIdentifier();
+            boolean isUseUserStoreDomainInLocalSubject = serviceProvider.getLocalAndOutBoundAuthenticationConfig()
+                    .isUseUserstoreDomainInLocalSubjectIdentifier();
+
+            if (StringUtils.isNotEmpty(sub)) {
+                // building subject in accordance with Local and Outbound Authentication Configuration preferences
+                if (isUseUserStoreDomainInLocalSubject) {
+                    sub = UserCoreUtil.addDomainToName(sub, userStoreDomain);
+                }
+                if (isUseTenantDomainInLocalSubject) {
+                    sub = UserCoreUtil.addTenantDomainToEntry(sub, userTenantDomain);
+                }
+            }
+        }
+        return sub;
+    }
+
+    private String getClientId(OAuth2TokenValidationResponseDTO tokenResponse) throws UserInfoEndpointException {
+        String clientId;
+        try {
+            clientId = OAuth2Util.getClientIdForAccessToken
```

Comment:
Can avoid local variable clientId by returning the statement from try block 

return OAuth2Util.getClientIdForAccessToken
                    (tokenResponse.getAuthorizationContextToken().getTokenString());